### DISCUSSION TITLE
Move most of safe polygon logic to @floating-ui/dom

### DISCUSSION THREAD

**Discussion by Etienne-M** [floating-ui/floating-ui#2867] [Ideas]
The base library would greatly benefit from the addition of safe polygon logic as a building block for developers to use.

For clarification, I did see the [discussion to add sub-menus](https://github.com/floating-ui/floating-ui/discussions/2344), but this is not what I am talking about. Let's leave all the
> list navigation, typeahead, portaling and getting it to work with screen readers 

to consumers of the library, and *only* provide safe polygon as a building block to use. This is to expand the use-cases covered by the base library, for example something simple like a tooltip with a scrollbar would benefit from this additional feature.

Looking at the [code](https://github.com/floating-ui/floating-ui/blob/master/packages/react/src/safePolygon.ts), it already seems like there are very few dependencies to react. I'm open to contributing if there's consensus.

**Comment by atomiks** [comment]
I'm hesitant to move any interaction logic to the vanilla library as it will broaden its scope beyond being a CSS anchor positioning polyfill, which is the most maintainable and clear to document.

Keeping interaction features limited to React helps clearly separate their roles - unless we wanted to port most other interactions to vanilla JS. It's hard to document which library supports what feature already.

**Comment by Etienne-M** [comment]
I completely understand the hesitation, it would be a nightmare to have to support complex interactions and components in the base library.

We're currently using the @floating-ui/dom library in a project that does not use react. So far, the high quality of this library enabled us to implement the components we need on our side (tooltip, combobox). This is the first time we're facing a "limitation", we have to add a tooltip that has potentially large content, and so we would limit the height of the content and add a scrollbar.

I could very well re-implement the safe polygon logic on our side, but this feels like a utility that could be directly in the core library. 

In the end, it is your decision

**Comment by Goldflow** [comment]
Late to the party but we're maintaining a web-component library, where we also use `@floating-ui/dom`-package for our  popover component.

I wanted to add another argument for adding this feature in the core library:

From 2026 onwards, we have stricter requirements on being WCAG compliant.

Related to tooltip's, we should be able to satisfy 1.4.13's hoverable criteria ([more info on 1.4.13 - Hoverable criteria here](https://www.w3.org/WAI/WCAG22/Understanding/content-on-hover-or-focus.html
))

having the safe polygon integrated in the `@floating-ui/dom`-package would be great for that reason as well

**Reply by AmmarGillani95** [reply]
We're also maintaining a web-component library - were you able to find a good solution to this?

**Reply by Goldflow** [reply]
for now, we've just added a minimum delay before it disappears
